chat/src/main/java/io/skygear/plugins/chat/RealmStore.java

```diff
@@ -0,0 +1,134 @@
+package io.skygear.plugins.chat;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import io.realm.OrderedRealmCollectionSnapshot;
+import io.realm.Realm;
+import io.realm.RealmConfiguration;
+import io.realm.RealmQuery;
+import io.realm.RealmResults;
+import io.realm.Sort;
+import io.realm.annotations.RealmModule;
+
+class RealmStore {
+
+    interface QueryBuilder<T> {
+        RealmQuery<T> buildQueryFrom(RealmQuery<T> baseQuery);
+    }
+
+    private final String name;
+    private final boolean inMemory;
+
+    RealmStore(String name, boolean inMemory) {
+        this.name = name;
+        this.inMemory = inMemory;
+    }
+
+    private Realm getRealm() {
+        RealmConfiguration.Builder configBuilder = new RealmConfiguration.Builder()
+                .name(this.name)
+                .schemaVersion(1)
+                .modules(new SkygearChatModule());
+
+        if (this.inMemory) {
+            configBuilder = configBuilder.inMemory();
+        }
+
+        return Realm.getInstance(configBuilder.build());
+    }
+
+    Message[] getMessages(QueryBuilder<MessageCacheObject> queryBuilder,
+                          int limit,
+                          String order) {
+        RealmQuery<MessageCacheObject> query = getRealm().where(MessageCacheObject.class);
+        query = queryBuilder.buildQueryFrom(query);
+        OrderedRealmCollectionSnapshot<MessageCacheObject> results = query.findAllSorted(order, Sort.DESCENDING).createSnapshot();
+        int size = results.size();
+        if (limit == -1 || limit > size) {
+            limit = size;
+        }
+
+        List<Message> messages = new ArrayList<>(limit);
+        List<MessageCacheObject> faultyCacheObjects = new ArrayList<>();
+        for (int i = 0; i < limit; i++) {
+            MessageCacheObject cacheObject = results.get(i);
+            if (cacheObject == null) {
+                throw new RuntimeException("query limit should not exceed result size");
```

Comment:
Doesn’t `results.get(i)` throws index out of bound exception? in this case the message in this exception seems wrong